05.wcコマンドを作るの解答提出

yokomaru opened the pull request:
### 必須要件
- [x] l、w、cの各オプションも実装すること。
- [x] 引数にファイル名が複数個来た場合にも対応し実行結果と出力を提出する。
- [x] 自作のls -lコマンドと自作のwcコマンドをつなげた出力を提出する。
      ruby ls.rb | ruby wc.rbのようにして組み合わせる。
- [x] 組み合わせた場合でもl、w、cの各オプションが使えるようにすること
- [x] 本物のls -lコマンド、本物のwcコマンドをつなげた出力を併記して提出すること。
- [x] rubocop-fjord のチェックが全てパスしたこともスクショで貼り付ける。

yokomaru committed:
add: main

mainメソッドの追加

yokomaru committed:
add: my-wc

yokomaru committed:
fix: rubocop and method name

rubocop指摘箇所修正
メソッド名が少しわかりづらかったのでgenerate_count_strings→generate_strings_for_countに変更
変数名lineがわかりづらかったので読み込んでいる行としてrowに変更

yoshitsugu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,63 @@
+    count << row.bytesize if options[:c]
+    counts << count
+    puts_count_string(count, args, string)
+  rescue SystemCallError
```
これだとシステムコールのエラーのときすべてここにきてしまいます。 `puts_error_message` 以外のシステムコールエラー時にも原理的にはここになってしまうので、これだと不適切です。ちゃんとディレクトリかどうか、もしくはそのファイルがあるかどうかはRubyレベルでチェックしてください。

yokomaru replied:
ありがとうございます！
ファイルチェックはFIle.file?やFile.directory?を使用して、その結果に応じてエラーメッセージを表示するようにいたします！

yoshitsugu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,63 @@
+def main
+  args = ARGV
+  options = parse_options(args)
+  strings = generate_strings_for_count(args)
```
`strings` だと名前が抽象的すぎます。変数名やメソッド名はその変数やメソッドが存在する意味となる目的や意図がわかるような名前にするように気をつけてください。

yoshitsugu replied:
また、15行目で結局 `args.empty? ? string : File.read(string)` としているので、 `strings_for_count` として期待されるカウント対象の文字列なっていないように思います。 `generate_strings_for_count` 内でカウント対象の文字列にしてしまうか、メソッド名を変えたほうがよいように思いました。

yoshitsugu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,63 @@
+  strings = generate_strings_for_count(args)
+  counts = []
+  strings.map do |string|
+    row = args.empty? ? string : File.read(string)
```
`row` は行列の行をあらわすときに使われる単語です。ここではあまりふさわしくないように思いました。

yoshitsugu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,63 @@
+  puts "#{display_rjust(count)} #{file_name}"
+end
+
+def count_total(counts, args)
```
ここでは `args.size > 1` のためにしか `args` が使われないですよね。引数のレベルで

```ruby
count_total(counts, multiple_arguments)
```

のように複数かどうかだけ渡すようにして、余計な情報を渡さないようにしたほうがよいと思います。関係ない情報までメソッドや関数に渡すことで関係ない部分に依存してしまうことをスタンプ結合などと呼び、結合度を測る指標においてこれも結合の一種と見做されます。(興味あればスタンプ結合で検索してみてください)

yokomaru replied:
>関係ない情報までメソッドや関数に渡すことで関係ない部分に依存してしまうことをスタンプ結合などと呼び、結合度を測る指標

スタンプ結合、とても勉強になりました！
メソッドが受け取った引数の内部的な構造を把握した実装になることも結合になるんですね・・とても納得しました。
今後は余計な情報を渡さず、必要な情報だけ渡すようにより意識いたします。

yoshitsugu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,63 @@
+
+def count_total(counts, args)
+  totals = counts.transpose.map(&:sum)
+  puts "#{display_rjust(totals)} total" if args.size > 1
```
`count_total` という名前からは合計をカウントするような動きを想定します。
その中で `puts` しているのはトリッキーであり、あまりメソッド名を見て使い方を想定できません。出力があるならそのような名前にしましょう。または値を返すだけにして、呼び出し側で `puts` するようにしましょう。

yokomaru replied:
>count_total という名前からは合計をカウントするような動きを想定します。
その中で puts しているのはトリッキーであり、あまりメソッド名を見て使い方を想定できません。出力があるならそのような名前にしましょう。

こちらは他のputs_count系の処理と統一し`puts_total_count`に修正いたしました！

yoshitsugu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,63 @@
+  puts "#{display_rjust(totals)} total" if args.size > 1
+end
+
+def display_rjust(array)
```
これもメソッド名から動きが想定できません。
もうちょっとメソッドの意図を書くようにしましょう。

yokomaru replied:
> これもメソッド名から動きが想定できません。

こちらはカウントのフォーマットの表示を調整したい意図でしたので
`adjust_format_display_count`のように修正いたしました！

今後はこちらを見直してメソッドの意図が伝わるように意識して命名いたします。
[名前重要！具体的過ぎるメソッド名や変数名の弊害と適切な抽象度 \| FJORD BOOT CAMP（フィヨルドブートキャンプ）](https://bootcamp.fjord.jp/articles/78)

yokomaru replied:
上記見直したところ動詞が3つになっているように見えてわかりづらかったので、シンプルに`format_count`に変更しました。

yoshitsugu reviewed (commented):
コメントしました。

yokomaru committed:
fix: method_name display_rjust

display_rjustという命名では動きを想像できないので、カウントの表示フォーマットを調整するという意味のadjust_format_display_countに変更

yokomaru committed:
fix: count_total

メソッド内でputsをしているのでputs_total_countにメソッド名を変更
 if args.size > 1 のチェックを外に出す
 counts.compactを追加

yokomaru committed:
fix: method name and design

- 抽象的になっていた各変数やメソッド名の変更
- 標準入力とファイル名を入力された時の処理を分ける(wc_standard_input, wc_input_files)
- SystemCallErrorでエラー処理していた箇所をファイルの存在チェック・ディレクトリチェックで分岐し、ハッシュにファイルに関する情報を保存するgenerate_file_contentに切り出す(ファイル名、ファイルの内容、ファイルタイプ、エラーメッセージ)
- count処理をメソッドに切り出す(count_input_text)

yokomaru committed:
add: puts_count_and_filename

標準入力時とファイル入力時でfilenameの表示の有無が異なるのでputs_countとputs_count_and_filenameでメソッドを分ける

yokomaru committed:
fix: File.exist? → File.file?

ディレクトリを指定したときにexist?でTrueになりFile.readでエラーになるのでFIle.file?チェックに変更

yokomaru committed:
delete: puts_count and puts_total_count

メソッド内でメソッドを読んでいるだけでメソッド化する意味があまりないので削除

yokomaru commented:
コメントありがとうございます。
指摘いただいた内容を踏まえ、一度設計から見直し修正を行いました！

## 修正前
- 入力の受け取りを標準入力とファイル名を同じメソッドで行っていた
   - そのため、stringsやrowといった変数名を使ってしまい全体的に抽象的な命名になっていた
   - 標準入力とファイル名で引数の有無が異なるため、どちらかの情報のみ表示したい時などにargs.empty?の分岐を各所で使う必要があった

## 修正後
- 標準入力とファイル名で処理を分ける(main処理でwc_standard_input, wc_input_filesに分岐する).  
    - 上記の処理は最初にarguments.empty?で分岐しているので他の箇所で分岐する必要がほぼなくなった
    - 変数名やメソッドなどに修正前より具体的な命名がつけやすくなった
    - ファイルのカウントなど本当に共通の処理のみ切り出す

まだ命名など自信がないため、再度レビューの程よろしくお願い致します！

yokomaru committed:
fix: parse_options

parse_optionsに渡す引数名の省略をやめてargumentsにする。

yokomaru committed:
add:  return array in wc_input_files else

ファイルタイプがfile以外の時でも複数引数に指定された時にtotalの計算結果を表示する必要があるため空文字をcount_input_textメソッドに渡して0の配列を作成する

yokomaru committed:
fix: rubocop

yokomaru committed:
fix: 不要なコメントを削除

yokomaru committed:
fix: method name

count_input_text → build_count
countする処理というよりcountという配列を作成する処理のため
generate_file_content → build_file_content
buildに統一
adjust_format_display_count → format_count
カウントをフォーマットするメソッドのためシンプルな命名に変更

yoshitsugu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,72 @@
+  options
+end
+
+def wc_standard_input(options)
+  input_source = $stdin.readlines.join
+  count = build_count(input_source, options)
+  puts format_count(count)
+end
+
+def wc_input_files(arguments, options)
+  counts = arguments.map do |argument|
+    file_content = build_file_content(argument)
+    if file_content[:type] == 'file'
+      count = build_count(file_content[:text], options)
+      puts "#{format_count(count)} #{file_content[:name]}"
+      count
+    else
+      puts file_content[:error_message]
+      build_count('', options) # 必要なオプション分の0の配列を返す
+    end
+  end
+  puts_total_count(counts) if arguments.size > 1
+end
```
発展的ですがこの辺もうちょっと統合できないか、考えてみてください。
標準入力からの入力をカウントする場合も、「ファイル名なし、ファイル内容が標準入力の内容」となっているファイルとして扱えば、後の処理は同じになりそうですよね。どうでしょうか。

yokomaru replied:
コメントありがとうございます！

>標準入力からの入力をカウントする場合も、「ファイル名なし、ファイル内容が標準入力の内容」となっているファイルとして扱えば、後の処理は同じ

なるほど、確かにそのように扱えば分ける必要がなさそうです・・！
再度修正し、wc_standard_inputとwc_input_filesで分けていた処理を統一してwcとしました。
また、合わせてあちこちでputsしていた処理をmainで一括でputsするように修正しました。

yoshitsugu reviewed (commented):
コメントしました。

yokomaru committed:
update: wc

- wc_standard_inputとwc_input_filesで分けていた処理を統合し、標準入力を名前が空のファイルとして扱う
- fileかどうかで分岐していた箇所をerror_messageの有無で分岐
- nameやerror_messageとともにcountの情報をhashに持たせる
- ところどころでputsしていた処理を修正し、再度にoutputsをputsする
- format_countにname情報もわたす

yokomaru committed:
update: parse_options

- optionsの分岐を三項演算子に変更

yokomaru committed:
fix: rubocop

yoshitsugu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,59 @@
+def main
+  arguments = ARGV
+  options = parse_options(arguments)
+  argument_empty = arguments.empty?
```
英語として `empty_argument` のほうが正しそうに思いました。もしくはbooleanっぽさをより出すなら `has_no_arguments` とかですかね。

yokomaru replied:
コメントありがとうございます！

> 英語として empty_argument のほうが正しそう

確かに形容詞＋名詞で正しい文法になってます・・！こちら今後気をつけます！

コメントいただいた ` has_no_arguments` の方が、
変数に何が入っているか分かりやすいと感じたのでこちらで修正いたします🙌

yoshitsugu reviewed (commented):
細かいところ一点だけコメントしました。

yokomaru committed:
update: change variable name

booleanを格納するargument_emptyという変数をhas_no_argumentsに変更

yoshitsugu reviewed (approved):
OKです！
